src/version_compare.rs

```diff
@@ -0,0 +1,443 @@
+use std::borrow::Cow;
+use std::cmp::Ordering;
+use std::fmt;
+
+
+#[derive(Debug, Eq, Default, Clone)]
+pub struct EVR<'a> {
+    epoch: Cow<'a, str>,
+    version: Cow<'a, str>,
+    release: Cow<'a, str>,
+}
+
+// TODO: sanity checking values
+impl<'a> EVR<'a> {
+    pub fn new<T: Into<Cow<'a, str>>>(epoch: T, version: T, release: T) -> EVR<'a> {
+        EVR {
+            epoch: epoch.into(),
+            version: version.into(),
+            release: release.into(),
+        }
+    }
+
+    pub fn epoch(&self) -> &str {
+        &self.epoch
+    }
+
+    pub fn version(&self) -> &str {
+        &self.version
+    }
+
+    pub fn release(&self) -> &str {
+        &self.release
+    }
+
+    pub fn values(&self) -> (&str, &str, &str) {
+        (&self.epoch, &self.version, &self.release)
+    }
+
+    pub fn parse_values(evr: &'a str) ->(&'a str, &'a str, &'a str) {
+        let (epoch, vr) = evr.split_once(':').unwrap_or(evr.split_at(0));
+        let (version, release) = vr.split_once('-').unwrap_or((vr, ""));
```

Comment:
Two ways to do this:

* No real validation on the inputs.  Almost never are you creating these from random values, they're coming from packages you already have.    `librpm` does this.  Nothing about the EVR comparison algorithm actually cares if there's a release or if the version is just a `.` so any restrictions we place on top of that would be basically artificial.
* We put some restrictions in place, like that `epoch` has to be only numeric values, release can't be empty, etc.